File: java/algorithm/binarysearch/leetcode/FindMinRotatedArray.java
package algorithm.binarysearch.leetcode;

/**
 * @author roseduan
 * @time 2022/2/9 10:20 下午
 * @description 搜索旋转排序数组中的最小值
 */
public class FindMinRotatedArray {
    /**
     * 线性查找
     */
    public int findMin1(int[] nums) {
        if (nums[0] <= nums[nums.length - 1]) {
            return nums[0];
        }
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < nums[i - 1]) {
                return nums[i];
            }
        }
        return nums[0];
    }

    /**
     * 二分查找
     */
    public int findMin2(int[] nums) {
        int low = 0, high = nums.length - 1;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] < nums[high]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return nums[low];
    }
}

File: cpp/search/leetcode/find_min_rotated_array.cpp
//
// Created by roseduan on 2022/2/12.
//

#include "iostream"
#include "vector"
using namespace std;

// 搜索旋转排序数组中的最小值
class Solution {
public:

    // 直接遍历查找
    int findMin1(vector<int>& nums) {
        int min = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] < nums[i - 1]) {
                return nums[i];
            }
        }
        return min;
    }

    // 二分查找
    int findMin2(vector<int>& nums) {
        int low = 0, high = nums.size() - 1;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] <= nums[high]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return nums[high];
    }
};
